libopae/plugins/xfpga/metrics/metrics_utils.c

```diff
@@ -293,40 +285,38 @@ fpga_result enum_powermgmt_metrics(fpga_metric_vector *vector,
 		return FPGA_INVALID_PARAM;
 	}
 
-	snprintf_s_ss(sysfs_path, sizeof(sysfs_path), "%s/%s", sysfspath, PWRMGMT);
-
-	dir = opendir(sysfs_path);
-	if (NULL == dir) {
-		FPGA_MSG("can't find dir %s ", strerror(errno));
+	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		FPGA_ERR("Failed pattern match %s: %s", sysfspath, strerror(errno));
+		globfree(&pglob);
 		return FPGA_NOT_FOUND;
 	}
 
-	while ((dirent = readdir(dir)) != NULL) {
-		if (!strcmp(dirent->d_name, "."))
-			continue;
-		if (!strcmp(dirent->d_name, ".."))
-			continue;
-		if (!strcmp(dirent->d_name, REVISION))
+	for (i = 0; i < pglob.gl_pathc; i++) {
+
+		char *dir_name = strrchr(pglob.gl_pathv[i], '/');
+
+		if (!dir_name)
 			continue;
 
-		snprintf_s_ss(metric_sysfs, sizeof(metric_sysfs), "%s/%s", sysfs_path, dirent->d_name);
+		if (!strcmp((dir_name + 1), REVISION))
+			continue;
 
-		result = get_metric_data_info(PWRMGMT, dirent->d_name, mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
+		result = get_metric_data_info(PWRMGMT, (dir_name + 1), mcp_metric_metadata, MCP_MDATA_SIZE, &metric_data);
 		if (result != FPGA_OK) {
 			FPGA_MSG("Failed to get metric metadata ");
 		}
 
-		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfs_path, dirent->d_name, metric_sysfs, metric_data.metric_units,
-									FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
+		result = add_metric_vector(vector, *metric_num, PWRMGMT, PWRMGMT, sysfspath, (dir_name + 1), pglob.gl_pathv[i], metric_data.metric_units,
+			FPGA_METRIC_DATATYPE_INT, FPGA_METRIC_TYPE_POWER, hw_type, 0);
 		if (result != FPGA_OK) {
 			FPGA_MSG("Failed to add metrics");
-			closedir(dir);
 			return result;
```

Comment:
do we clean up the vector outside of this? even on error?

Reply:
yes
Enum Metrics cleans vector